examples/aggregator/module/rust/src/data.rs

```diff
@@ -0,0 +1,74 @@
+//
+// Copyright 2020 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::proto::aggregator::SerializedSparseVector;
+use aggregator_common::Monoid;
+use std::collections::HashMap;
+use std::convert::{From, TryFrom};
+
+// Sparce Vectors are stored as Hash Maps.
+#[derive(Clone, Debug)]
+pub struct SparseVector {
+    map: HashMap<u32, f32>,
+}
+
+// Deserializes a Protobuf message into an internal Sparce Vector implementation.
+// If a Protobuf message has duplicated indices, returns an `Error`.
+impl TryFrom<SerializedSparseVector> for SparseVector {
+    type Error = String;
+
+    fn try_from(src: SerializedSparseVector) -> Result<Self, Self::Error> {
+        src.indices.iter().zip(src.values.iter()).try_fold(
+            SparseVector::identity(),
+            |mut svec, (&i, &v)| match svec.map.get(&i) {
+                Some(_) => Err(format!("Duplicated index: {}", i)),
+                None => Ok({
+                    svec.map.insert(i, v);
```

Comment:
But even if I will use `entry` instead of `insert` I will still need to reborrow the map in the `fold`.
Or you mean not using `fold` at all?

Reply:
I meant `entry` instead of `get` actually, and keep the `fold`. I'm not even sure if it works though, let me know if you try it out.

Reply:
The problem with `entry` is that I still need to return error from the function inside `try_fold`.
Also, I'm not sure, why `Copy` is necessary - `svec` in the `|mut svec, (&i, &v)|` is just moved.

Reply:
To clarify: I am not talking about `svec`; I think you can currently re-borrow map inside match, because when you do `match sve.map.get(&i)`, this value is immediately copied, and therefore the borrow from the `get` does not last inside the body of the `match`. If it did (i.e. if the type was not `Copy`, then I think you would not be able to call `insert` here. But I haven't actually tried :)

Reply:
Done

Reply:
But I'm returning `Ok(svec)` from the `Entry::Vacant(entry)`.
Does it mean that I'm still reborrowing `svec` mutably?

Reply:
I haven't actually tried any of the stuff I mentioned here, it's just something that occurred to me :) and I was not necessarily suggesting anything needs to change in the code. I was mostly hoping you could try out various options (e.g. Copy vs non-Copy, entry vs get) and let the rest of us know :)

Reply:
It may well be that the only way to make this work for a non-Copy type is to just use a `for` loop and `entry` to access an individual entry from the map. Let me know if you end up trying that out, separately. I think the code here looks fine anyways, no need to change it further, unless you want to.